dali/core/mm/async_pool_test.cu

```diff
@@ -0,0 +1,325 @@
+  mm::test::test_device_resource upstream;
+  {
+    async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+
+    vector<std::thread> threads;
+
+    for (int t = 0; t < 10; t++) {
+      threads.push_back(std::thread([&]() {
+        CUDAStream stream = CUDAStream::Create(true);
+        vector<block> blocks;
+        detail::dummy_lock mtx;
+        AsyncPoolTest(pool, blocks, mtx, stream);
+        CUDA_CALL(cudaStreamSynchronize(stream));
+      }));
+    }
+    for (auto &t : threads)
+      t.join();
+  }
+  std::cerr << "Peak consumption:     " << upstream.get_peak_size() << " bytes\n";
+  std::cerr << "Upstream allocations: " << upstream.get_num_allocs() << std::endl;
+  upstream.check_leaks();
+}
+
+
+TEST(MMAsyncPool, TwoStream) {
+  mm::test::test_device_resource upstream;
+  CUDAStream s1 = CUDAStream::Create(true);
+  CUDAStream s2 = CUDAStream::Create(true);
+  stream_view sv1(s1);
+  stream_view sv2(s2);
+
+  GPUHog hog;
+  hog.init();
+  const int min_success = 10;
+  const int max_not_busy = 100;
+  int stream_not_busy = 0;
+  int success = 0;
+  while (success < min_success) {
+    async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+    void *p1 = pool.allocate_async(1000, sv1);
+    hog.run(s1);
+    pool.deallocate_async(p1, 1000, sv1);
+    void *p2 = pool.allocate_async(1000, sv2);
+    void *p3 = pool.allocate_async(1000, sv1);
+    cudaError_t e = cudaStreamQuery(s1);
+    if (e != cudaErrorNotReady) {
+      std::cerr << "Stream s1 finished before attempt to allocate on s2 was made - retrying\n";
+      CUDA_CALL(cudaGetLastError());
+      if (++stream_not_busy > max_not_busy) {
+        FAIL() << "Stream s1 finished - test unreliable.";
+      }
+      continue;
+    }
+    stream_not_busy = 0;
+    ASSERT_NE(p1, p2);
+    ASSERT_EQ(p1, p3);
+    cudaStreamSynchronize(s1);
+    success++;
+    cudaStreamSynchronize(s2);
+  }
+  std::cerr << "Peak consumption:     " << upstream.get_peak_size() << " bytes\n";
+  std::cerr << "Upstream allocations: " << upstream.get_num_allocs() << std::endl;
+  upstream.check_leaks();
+}
+
+TEST(MMAsyncPool, MultiStreamRandomWithGPUHogs) {
+  mm::test::test_device_resource upstream;
+  {
+    async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+
+    vector<std::thread> threads;
+
+    for (int t = 0; t < 10; t++) {
+      threads.push_back(std::thread([&]() {
+        CUDAStream stream = CUDAStream::Create(true);
+        vector<block> blocks;
+        detail::dummy_lock mtx;
+        AsyncPoolTest(pool, blocks, mtx, stream, 20000, true);
+        CUDA_CALL(cudaStreamSynchronize(stream));
+      }));
+    }
+    for (auto &t : threads)
+      t.join();
+  }
+  std::cerr << "Peak consumption:     " << upstream.get_peak_size() << " bytes\n";
+  std::cerr << "Upstream allocations: " << upstream.get_num_allocs() << std::endl;
+  upstream.check_leaks();
+}
+
+
+TEST(MMAsyncPool, CrossStream) {
```

Comment:
Does it make any sense to add `CrossStream` test with Hogs ?

Reply:
I could.

Reply:
Done.